### PR TITLE
IGNITE-11922: Migration of ClusterMetricsMXBeanImpl to the new metric framework

### PR DESCRIPTION
**Pull request by 1vanan**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java [hunk 1]
```diff
@@ -36,6 +39,12 @@
     /** Job metrics update time. */
     private volatile long jobMetricsUpdateTime;
 
+    /** System metrics prefix. */
+    public static final String SYS_METRICS = "sys";
+
+    /** System metrics. */
+    MetricRegistry sysreg;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java]: This variable is not necessary in the class.

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java [hunk 2]
```diff
@@ -295,7 +306,10 @@ public ClusterMetricsImpl(GridKernalContext ctx, GridLocalMetrics vmMetrics, lon
 
     /** {@inheritDoc} */
     @Override public long getNodeStartTime() {
-        return nodeStartTime;
+        LongMetricImpl startTime = sysreg.metric("NodeStartTime", null);
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java]: Why *NodeStartTime* changed during it's request?

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java [hunk 3]
```diff
@@ -350,15 +364,23 @@ public ClusterMetricsImpl(GridKernalContext ctx, GridLocalMetrics vmMetrics, lon
 
     /** {@inheritDoc} */
     @Override public int getTotalNodes() {
-        return 1;
+        IntMetricImpl totalNodes = sysreg.intMetric("TotalNodes", null);
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java]: Why this change? Please, revert it.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java [hunk 4]
```diff
@@ -136,6 +146,10 @@ public int sentMessagesCount() {
         if (res0 < 0)
             res0 = Integer.MAX_VALUE;
 
+
+        IntMetricImpl msgcnt = mreg.intMetric("NodeStartTime", null);
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: NodeStartTime?

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java [hunk 5]
```diff
@@ -48,6 +51,9 @@
     /** Sent bytes count.*/
     private final LongAdder sentBytesCnt = new LongAdder();
 
+    /** Metric registry. */
+    MetricRegistry mreg;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: We should cache specific counters in the constructor instead of querying it from a registry on each access.

**Comment by nizhikov** [comment]
1. Please, solve merge conflicts
2. We should cache specific Counters instead of query it from registry on each access.
3. Please, double check usages of `add` and `set`. `add` - ads some value to the counter while `set` sets it. Seems, we have issues with it in the PR.

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java [hunk 6]
```diff
@@ -350,15 +364,23 @@ public ClusterMetricsImpl(GridKernalContext ctx, GridLocalMetrics vmMetrics, lon
...
+
+        return totalNodes.value();
     }
 
     /** {@inheritDoc} */
     @Override public long getCurrentPmeDuration() {
         GridDhtPartitionsExchangeFuture future = ctx.cache().context().exchange().lastTopologyFuture();
-        
-        return (future == null || future.isDone()) ?
+
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java]: Seems, we should change `PmeStartTime` on `GridDhtPartitionsExchangeFuture` start instead of calculation on query.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java [hunk 7]
```diff
@@ -394,6 +397,10 @@ public GridDhtPartitionsExchangeFuture(
             }
         };
 
+        MetricRegistry sysreg = cctx.kernalContext().metric().registry(SYS_METRICS);
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java]: Can we pass metric via constructor or use `findMetric` here instead of trying to create it on each PME?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskProcessor.java [hunk 8]
```diff
@@ -122,8 +124,8 @@
     /** */
     private final GridLocalEventListener discoLsnr;
 
-    /** Total executed tasks. */
-    private final LongAdder execTasks = new LongAdder();
+    /** Total executed tasks metric. */
+    IntMetricImpl executedTasks;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskProcessor.java]: Let's use LongMetricImpl here.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/version/GridCacheVersionManager.java [hunk 9]
```diff
@@ -82,10 +87,20 @@
         }
     };
 
+    /**
+     * Default constructor.
+     */
+    public GridCacheVersionManager() {
+        MetricRegistry sysreg = cctx.kernalContext().metric().registry("cache");
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/version/GridCacheVersionManager.java]: Let's use `CacheMetricsImpl#CACHE_METRICS_PREFIX` constant here.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java [hunk 10]
```diff
@@ -406,6 +408,9 @@
     private boolean enableTroubleshootingLog = IgniteSystemProperties
         .getBoolean(IgniteSystemProperties.IGNITE_TROUBLESHOOTING_LOGGER);
 
+    /** Io metric. */
+    private MetricRegistry ioMetric = ((IgniteKernal)ignite).context().metric().registry("io");
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java]: IgniteKernal -> IgniteEx

File: modules/core/src/main/java/org/apache/ignite/internal/managers/communication/GridIoManager.java [hunk 11]
```diff
@@ -139,6 +142,21 @@
     /** Direct protocol version. */
     public static final byte DIRECT_PROTO_VER = 3;
 
+    /** Sent messages count metric. */
+    private final IntGauge sentMessagesCnt;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/communication/GridIoManager.java]: According to abbrevation list it should be - `sentMessagesCnt` -> `sentMsgCnt`

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java [hunk 12]
```diff
@@ -491,10 +528,10 @@ public ClusterMetricsImpl(GridKernalContext ctx, long nodeStartTime) {
 
     /** {@inheritDoc} */
     @Override public long getCurrentPmeDuration() {
-        GridDhtPartitionsExchangeFuture future = ctx.cache().context().exchange().lastTopologyFuture();
-        
-        return (future == null || future.isDone()) ?
-            0 : TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - future.getStartTime());
+        GridDhtPartitionsExchangeFuture fut = ctx.cache().context().exchange().lastTopologyFuture();
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java]: Why this change?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheSharedManagerAdapter.java [hunk 13]
```diff
@@ -58,6 +65,9 @@
 
         diagnosticLog = cctx.logger(DIAGNOSTIC_LOG_CATEGORY);
 
+        MetricRegistry sysreg = cctx.kernalContext().metric().registry(CACHE_METRICS);
+        lastDataVer = sysreg.longMetric("LastDataVersion", "Node start time.");
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheSharedManagerAdapter.java]: Please, add empty line.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskProcessor.java [hunk 14]
```diff
@@ -122,8 +124,8 @@
     /** */
     private final GridLocalEventListener discoLsnr;
 
-    /** Total executed tasks. */
-    private final LongAdder execTasks = new LongAdder();
+    /** Total executed tasks metric. */
+    LongAdderMetric executedTasks;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskProcessor.java]: Let's keep name `execTasks`?

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java [hunk 15]
```diff
@@ -1143,6 +1149,14 @@ private void connectedNew(
     /** NIO server. */
     private GridNioServer<Message> nioSrvr;
 
+    /** Outbound messages queue size metric. */
+    private final IntGauge outBoundMsg = new IntGauge("OutboundMessagesQueueSize", null,
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java]: This metric doesn't register.

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/ApiParity/TcpCommunicationSpiParityTest.cs [hunk 16]
```diff
@@ -61,7 +61,8 @@ public class TcpCommunicationSpiParityTest
             "endTime",
             "id",
             "connectionIndex",
-            "NodeFilter"
+            "NodeFilter",
```
Comment (nizhikov) [inline modules/platforms/dotnet/Apache.Ignite.Core.Tests/ApiParity/TcpCommunicationSpiParityTest.cs]: Why this change?

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java [hunk 17]
```diff
@@ -193,6 +218,18 @@ public ClusterMetricsImpl(GridKernalContext ctx, long nodeStartTime) {
         nonHeapUsed = mreg.findMetric(metricName("memory", "nonheap", "used"));
         nonHeapCommitted = mreg.findMetric(metricName("memory", "nonheap", "committed"));
         nonHeapMax = mreg.findMetric(metricName("memory", "nonheap", "max"));
+
+        MetricRegistry cacheReg = ctx.metric().registry("cache");
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java]: Let's remove `cacheReg` variable.

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/ApiParity/TcpCommunicationSpiParityTest.cs (hunk 16 above)
Comment (1vanan) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/ApiParity/TcpCommunicationSpiParityTest.cs]: Because without such change some .NET [1] tests will be failed.
[1] https://ci.ignite.apache.org/viewLog.html?buildId=4387853&buildTypeId=IgniteTests24Java8_RunAll#testNameId5475589648862867137

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java [hunk 18]
```diff
@@ -181,6 +207,7 @@ public ClusterMetricsImpl(GridKernalContext ctx, long nodeStartTime) {
         peakThreadCnt = mreg.findMetric(PEAK_THREAD_CNT);
         totalStartedThreadCnt = mreg.findMetric(TOTAL_STARTED_THREAD_CNT);
         daemonThreadCnt = mreg.findMetric(DAEMON_THREAD_CNT);
+        executedTasks = mreg.findMetric("TotalExecutedTasks");
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java]: Let's make "TotalExecutedTasks" and other metric names constant in `GridIoManager`

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/ApiParity/TcpCommunicationSpiParityTest.cs (hunk 16 above)
Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/ApiParity/TcpCommunicationSpiParityTest.cs]: Seems, it should be reverted.
